File: lua-lib/aiomsg/ws.lua
-- Server-side WebSocket (RFC 6455) byte-stream adapter for aiomsg bind sockets
-- (see ../PROTOCOL.md §10).
--
-- This module is the Lua counterpart of python-lib/aiomsg/ws.py. It is pure
-- byte manipulation — no sockets — so it slots into the reactor in socket.lua
-- and is unit-testable in isolation. It implements only the server half of
-- RFC 6455 and never negotiates any subprotocol or extension.
--
-- The two public pieces the reactor needs:
--   * the HTTP upgrade (parse_upgrade / success_response / error_response), and
--   * the frame layer: encode() to wrap one aiomsg frame as a binary WebSocket
--     message, and a parser (new_parser + ingest) that turns the concatenation
--     of inbound binary-message payloads back into the aiomsg byte stream of
--     §2, disposing of interleaved control frames (ping->pong, pong ignored,
--     close echoed). WebSocket message boundaries carry no meaning.
--
-- Like the rest of aiomsg (see protocol.lua) this stays portable across Lua 5.1,
-- LuaJIT and 5.3+ by doing its own arithmetic bit-twiddling rather than relying
-- on the native bitwise operators, which only 5.3+ can even parse. SHA-1 is
-- hand-rolled here so the dependency set stays exactly LuaSocket + LuaSec (LuaSec
-- exposes no general digest); base64 is LuaSocket's mime.

local mime = require("mime")

local M = {}

M.GUID = "258EAFA5-E914-47DA-95CA-C5AB0DC85B11"

-- Opcodes (RFC 6455 §5.2).
M.OP_CONT = 0x0
M.OP_TEXT = 0x1
M.OP_BINARY = 0x2
M.OP_CLOSE = 0x8
M.OP_PING = 0x9
M.OP_PONG = 0xA

M.MAX_REQUEST_BYTES = 8192 -- cap the pre-upgrade HTTP request

-- --- Portable 32-bit bitwise ops --------------------------------------------
--
-- Bit-by-bit over Lua numbers, so this module parses and runs on 5.1/LuaJIT as
-- well as 5.3+ (see the note at the top). Operands and results are unsigned
-- 32-bit values; every intermediate stays below 2^32, well within a double's
-- exact-integer range. Only SHA-1 needs the full and/or/xor; the frame layer
-- gets by with plain arithmetic.

local function band(a, b)
  local r, bit = 0, 1
  while a > 0 and b > 0 do
    if a % 2 == 1 and b % 2 == 1 then
      r = r + bit
    end
    a, b, bit = math.floor(a / 2), math.floor(b / 2), bit * 2
  end
  return r
end

local function bor(a, b)
  local r, bit = 0, 1
  while a > 0 or b > 0 do
    if a % 2 == 1 or b % 2 == 1 then
      r = r + bit
    end
    a, b, bit = math.floor(a / 2), math.floor(b / 2), bit * 2
  end
  return r
end

local function bxor(a, b)
  local r, bit = 0, 1
  while a > 0 or b > 0 do
    if a % 2 ~= b % 2 then
      r = r + bit
    end
    a, b, bit = math.floor(a / 2), math.floor(b / 2), bit * 2
  end
  return r
end

-- Left-rotate a 32-bit word by n bits (0 < n < 32). The shifted-out high bits
-- and the surviving low bits land in disjoint positions, so a plain add rebuilds
-- the word, and neither term ever reaches 2^32.
local function rotl32(x, n)
  x = x % 0x100000000
  local hi = 2 ^ (32 - n)
  return (x % hi) * (2 ^ n) + math.floor(x / hi)
end

-- --- SHA-1 (hand-rolled: LuaSec exposes no digest) ---------------------------

-- SHA-1 digest of a byte string, returned as 20 raw bytes.
local function sha1(msg)
  local h0, h1, h2, h3, h4 = 0x67452301, 0xEFCDAB89, 0x98BADCFE, 0x10325476, 0xC3D2E1F0
  local ml = #msg * 8
  msg = msg .. "\128"
  while #msg % 64 ~= 56 do
    msg = msg .. "\0"
  end
  local hi = math.floor(ml / 4294967296) % 4294967296
  local lo = ml % 4294967296
  msg = msg
    .. string.char(math.floor(hi / 16777216) % 256, math.floor(hi / 65536) % 256,
                   math.floor(hi / 256) % 256, hi % 256,
                   math.floor(lo / 16777216) % 256, math.floor(lo / 65536) % 256,
                   math.floor(lo / 256) % 256, lo % 256)
  for chunk = 1, #msg, 64 do
    local w = {}
    for i = 0, 15 do
      local a, b, c, d = string.byte(msg, chunk + i * 4, chunk + i * 4 + 3)
      w[i] = a * 16777216 + b * 65536 + c * 256 + d
    end
    for i = 16, 79 do
      w[i] = rotl32(bxor(bxor(bxor(w[i - 3], w[i - 8]), w[i - 14]), w[i - 16]), 1)
    end
    local a, b, c, d, e = h0, h1, h2, h3, h4
    for i = 0, 79 do
      local f, k
      if i < 20 then
        f = bor(band(b, c), band(0xFFFFFFFF - b, d))
        k = 0x5A827999
      elseif i < 40 then
        f = bxor(bxor(b, c), d)
        k = 0x6ED9EBA1
      elseif i < 60 then
        f = bor(bor(band(b, c), band(b, d)), band(c, d))
        k = 0x8F1BBCDC
      else
        f = bxor(bxor(b, c), d)
        k = 0xCA62C1D6
      end
      local temp = (rotl32(a, 5) + f + e + k + w[i]) % 0x100000000
      e, d, c, b, a = d, c, rotl32(b, 30), a, temp
    end
    h0 = (h0 + a) % 0x100000000
    h1 = (h1 + b) % 0x100000000
    h2 = (h2 + c) % 0x100000000
    h3 = (h3 + d) % 0x100000000
    h4 = (h4 + e) % 0x100000000
  end
  local function bytes(h)
    return string.char(math.floor(h / 16777216) % 256, math.floor(h / 65536) % 256,
                       math.floor(h / 256) % 256, h % 256)
  end
  return bytes(h0) .. bytes(h1) .. bytes(h2) .. bytes(h3) .. bytes(h4)
end

-- The Sec-WebSocket-Accept value for a client key (RFC 6455 §4.2.2):
-- base64(SHA1(key .. GUID)). Test vector: dGhlIHNhbXBsZSBub25jZQ== ->
-- s3pPLMBiTxaQ9kYGzzhZRbK+xOo=.
function M.compute_accept(key)
  return (mime.b64(sha1(key .. M.GUID)))
end

-- --- HTTP upgrade ------------------------------------------------------------

-- Validate an HTTP upgrade request (bytes through the blank line). Returns the
-- Sec-WebSocket-Key on success, or nil plus the HTTP status line to send back on
-- any violation. Host/Origin/Sec-WebSocket-Protocol/-Extensions are ignored: no
-- subprotocol or extension is negotiated, and auth is out of scope (§1.5).
function M.parse_upgrade(request)
  local lines = {}
  for line in (request .. "\r\n"):gmatch("(.-)\r\n") do
    lines[#lines + 1] = line
  end
  local method, _, version = (lines[1] or ""):match("^(%S+)%s+(%S+)%s+(%S+)$")
  if method ~= "GET" or not (version and version:match("^HTTP/1%.")) then
    return nil, "400 Bad Request"
  end
  local headers = {}
  for i = 2, #lines do
    local name, value = lines[i]:match("^([^:]+):%s*(.*)$")
    if name then
      headers[name:lower()] = value
    end
  end
  if (headers["upgrade"] or ""):lower() ~= "websocket" then
    return nil, "400 Bad Request"
  end
  local has_upgrade = false
  for token in (headers["connection"] or ""):gmatch("[^,]+") do
    if token:gsub("%s", ""):lower() == "upgrade" then
      has_upgrade = true
    end
  end
  if not has_upgrade then
    return nil, "400 Bad Request"
  end
  if (headers["sec-websocket-version"] or "") ~= "13" then
    return nil, "426 Upgrade Required"
  end
  local key = headers["sec-websocket-key"] or ""
  local decoded = mime.unb64(key)
  if not decoded or #decoded ~= 16 then
    return nil, "400 Bad Request"
  end
  return key
end

function M.success_response(key)
  return "HTTP/1.1 101 Switching Protocols\r\n"
    .. "Upgrade: websocket\r\n"
    .. "Connection: Upgrade\r\n"
    .. "Sec-WebSocket-Accept: " .. M.compute_accept(key) .. "\r\n"
    .. "\r\n"
end

function M.error_response(status)
  if status:sub(1, 3) == "426" then
    return "HTTP/1.1 " .. status .. "\r\nSec-WebSocket-Version: 13\r\n\r\n"
  end
  return "HTTP/1.1 " .. status .. "\r\n\r\n"
end

-- --- Frame layer -------------------------------------------------------------

-- Encode one server->client frame: FIN=1, RSV=0, unmasked, given opcode.
function M.encode(opcode, payload)
  local n = #payload
  local header
  -- TODO(frame-size): enforce a configurable maximum frame length
  if n < 126 then
    header = string.char(0x80 + opcode, n)
  elseif n < 65536 then
    header = string.char(0x80 + opcode, 126, math.floor(n / 256) % 256, n % 256)
  else
    header = string.char(0x80 + opcode, 127,
      math.floor(n / 2 ^ 56) % 256, math.floor(n / 2 ^ 48) % 256,
      math.floor(n / 2 ^ 40) % 256, math.floor(n / 2 ^ 32) % 256,
      math.floor(n / 2 ^ 24) % 256, math.floor(n / 2 ^ 16) % 256,
      math.floor(n / 256) % 256, n % 256)
  end
  return header .. payload
end

local function u16(code)
  return string.char(math.floor(code / 256) % 256, code % 256)
end

-- Try to parse one masked client frame from `buf`.
-- Returns (frame, consumed) on a complete frame, (nil) if more bytes are
-- needed, or (nil, nil, code) on a protocol error (close with `code`).
local function parse_frame(buf)
  if #buf < 2 then
    return nil
  end
  local b0, b1 = string.byte(buf, 1, 2)
  if math.floor(b0 / 16) % 8 ~= 0 then -- any RSV bit set
    return nil, nil, 1002
  end
  local fin = math.floor(b0 / 128) % 2
  local opcode = b0 % 16
  if math.floor(b1 / 128) % 2 ~= 1 then -- every client frame MUST be masked
    return nil, nil, 1002
  end
  local len = b1 % 128
  local offset = 2
  if len == 126 then
    if #buf < 4 then
      return nil
    end
    local a, b = string.byte(buf, 3, 4)
    len = a * 256 + b
    offset = 4
  elseif len == 127 then
    if #buf < 10 then
      return nil
    end
    -- TODO(frame-size): enforce a configurable maximum frame length
    local b3 = string.byte(buf, 3)
    if math.floor(b3 / 128) % 2 == 1 then -- 64-bit length MSB MUST be 0
      return nil, nil, 1002
    end
    len = 0
    for i = 3, 10 do
      len = len * 256 + string.byte(buf, i)
    end
    offset = 10
  end
  if opcode >= 0x8 and (fin ~= 1 or len > 125) then -- control frames: FIN=1, <=125
    return nil, nil, 1002
  end
  if #buf < offset + 4 + len then
    return nil
  end
  local mask = { string.byte(buf, offset + 1, offset + 4) }
  local masked = string.sub(buf, offset + 5, offset + 4 + len)
  local out = {}
  for i = 1, len do
    out[i] = string.char(bxor(string.byte(masked, i), mask[(i - 1) % 4 + 1]))
  end
  return { opcode = opcode, fin = fin, payload = table.concat(out) }, offset + 4 + len
end

-- A parser accumulates raw bytes across reads (WS frames may be split).
function M.new_parser()
  return { buf = "" }
end

-- Consume `raw` bytes. Returns:
--   payload : concatenation of binary/continuation payloads (the aiomsg stream);
--   control : bytes to send back to the client (pong and/or a close frame);
--   closed  : true if the connection must be torn down after flushing.
function M.ingest(state, raw)
  state.buf = state.buf .. raw
  local payload, control = {}, {}
  local closed = false
  while true do
    local frame, consumed, err = parse_frame(state.buf)
    if err then
      control[#control + 1] = M.encode(M.OP_CLOSE, u16(err))
      closed = true
      break
    end
    if not frame then
      break -- need more bytes
    end
    state.buf = string.sub(state.buf, consumed + 1)
    local op = frame.opcode
    if op == M.OP_BINARY or op == M.OP_CONT then
      payload[#payload + 1] = frame.payload
    elseif op == M.OP_PING then
      control[#control + 1] = M.encode(M.OP_PONG, frame.payload)
    elseif op == M.OP_PONG then
      -- ignore
    elseif op == M.OP_TEXT then
      control[#control + 1] = M.encode(M.OP_CLOSE, u16(1003)) -- text is not valid for aiomsg
      closed = true
      break
    elseif op == M.OP_CLOSE then
      local code = #frame.payload >= 2 and (string.byte(frame.payload, 1) * 256 + string.byte(frame.payload, 2)) or 1000
      control[#control + 1] = M.encode(M.OP_CLOSE, u16(code))
      closed = true
      break
    else
      control[#control + 1] = M.encode(M.OP_CLOSE, u16(1002)) -- unknown opcode
      closed = true
      break
    end
  end
  return table.concat(payload), table.concat(control), closed
end

return M

File: python-lib-blocking/aiomsg_blocking/ws.py
"""
aiomsg_blocking.ws
==================

Server-side WebSocket (RFC 6455) byte-stream adapter for aiomsg **bind** sockets
(see ``PROTOCOL.md`` §10). The blocking-socket counterpart of
``python-lib/aiomsg/ws.py``.

The single public entry point is :func:`sniff_and_wrap`. Given a freshly accepted
(and TLS-terminated, if any) ``socket.socket``/``ssl.SSLSocket`` it peeks the
first byte, decides whether the peer speaks raw aiomsg or HTTP, performs the
WebSocket upgrade when needed, and returns a **socket-like object** that the
existing framing layer (:mod:`aiomsg_blocking.msgproto`) drives unchanged — it
only ever calls ``recv_into`` / ``send`` / ``settimeout`` / ``fileno`` /
``pending`` on the socket, and both wrappers here provide exactly those.

WebSocket message boundaries carry no meaning: ``recv_into`` yields the
concatenation of all inbound binary-message payloads (the byte stream of §2), and
each aiomsg ``send`` goes out as one unmasked binary frame. Only the server half
of RFC 6455 is implemented, and no subprotocol or extension is negotiated.
"""

import base64
import hashlib
import logging
import ssl

logger = logging.getLogger(__name__)

WS_GUID = "258EAFA5-E914-47DA-95CA-C5AB0DC85B11"

# Opcodes (RFC 6455 §5.2).
OP_CONT = 0x0
OP_TEXT = 0x1
OP_BIN = 0x2
OP_CLOSE = 0x8
OP_PING = 0x9
OP_PONG = 0xA

MAX_REQUEST_BYTES = 8192
HANDSHAKE_TIMEOUT = 15.0  # deadline for the sniff + upgrade exchange (§8)

_SEND_RETRYABLE = (TimeoutError, ssl.SSLWantReadError, ssl.SSLWantWriteError)


class WSError(Exception):
    """A rejected upgrade request; ``str(self)`` is the HTTP status line."""


class _ProtocolError(Exception):
    """A fatal WebSocket framing violation; ``code`` is the close code to send."""

    def __init__(self, code: int):
        super().__init__(f"WebSocket protocol error {code}")
        self.code = code


# --- Handshake (pure) -------------------------------------------------------


def compute_accept(key: str) -> str:
    """base64(SHA1(key + GUID)) per RFC 6455 §4.2.2. Vector:
    ``dGhlIHNhbXBsZSBub25jZQ==`` → ``s3pPLMBiTxaQ9kYGzzhZRbK+xOo=``."""
    digest = hashlib.sha1((key + WS_GUID).encode("ascii")).digest()
    return base64.b64encode(digest).decode("ascii")


def parse_upgrade(request: bytes) -> str:
    """Validate the HTTP upgrade request bytes and return its
    ``Sec-WebSocket-Key``. Raises :class:`WSError` with the HTTP status line on
    any violation. Host/Origin/subprotocol/extensions are ignored (§1.5)."""
    lines = request.split(b"\r\n")
    parts = lines[0].split(b" ")
    if len(parts) != 3 or parts[0] != b"GET" or not parts[2].startswith(b"HTTP/1."):
        raise WSError("400 Bad Request")

    headers = {}
    for line in lines[1:]:
        if not line:
            continue
        name, sep, value = line.partition(b":")
        if not sep:
            raise WSError("400 Bad Request")
        headers[name.strip().lower()] = value.strip()

    if headers.get(b"upgrade", b"").lower() != b"websocket":
        raise WSError("400 Bad Request")
    tokens = [t.strip().lower() for t in headers.get(b"connection", b"").split(b",")]
    if b"upgrade" not in tokens:
        raise WSError("400 Bad Request")
    if headers.get(b"sec-websocket-version", b"") != b"13":
        raise WSError("426 Upgrade Required")

    key = headers.get(b"sec-websocket-key", b"")
    try:
        if len(base64.b64decode(key, validate=True)) != 16:
            raise ValueError
    except Exception:
        raise WSError("400 Bad Request")
    return key.decode("ascii")


def success_response(key: str) -> bytes:
    return (
        "HTTP/1.1 101 Switching Protocols\r\n"
        "Upgrade: websocket\r\n"
        "Connection: Upgrade\r\n"
        f"Sec-WebSocket-Accept: {compute_accept(key)}\r\n"
        "\r\n"
    ).encode("ascii")


def error_response(status: str) -> bytes:
    if status.startswith("426"):
        return f"HTTP/1.1 {status}\r\nSec-WebSocket-Version: 13\r\n\r\n".encode("ascii")
    return f"HTTP/1.1 {status}\r\n\r\n".encode("ascii")


# --- Frame layer (pure) -----------------------------------------------------


def server_frame(opcode: int, payload: bytes) -> bytes:
    """Encode one server→client frame: FIN=1, RSV=0, unmasked, given opcode."""
    n = len(payload)
    header = bytearray((0x80 | opcode,))
    # TODO(frame-size): enforce a configurable maximum frame length
    if n < 126:
        header.append(n)
    elif n < 65536:
        header.append(126)
        header += n.to_bytes(2, "big")
    else:
        header.append(127)
        header += n.to_bytes(8, "big")
    return bytes(header) + payload


def parse_client_frame(buf: bytes):
    """Try to parse one masked client frame from the front of ``buf``.

    Returns ``(opcode, payload, consumed)`` for a complete frame, or ``None`` if
    more bytes are needed. Raises :class:`_ProtocolError` on a fatal violation
    (unmasked frame, RSV bits, bad control frame, 64-bit length MSB set).
    """
    if len(buf) < 2:
        return None
    b0, b1 = buf[0], buf[1]
    if b0 & 0x70:  # any RSV bit set
        raise _ProtocolError(1002)
    opcode = b0 & 0x0F
    fin = bool(b0 & 0x80)
    if not (b1 & 0x80):  # every client frame MUST be masked
        raise _ProtocolError(1002)
    length = b1 & 0x7F
    offset = 2
    if length == 126:
        if len(buf) < 4:
            return None
        length = int.from_bytes(buf[2:4], "big")
        offset = 4
    elif length == 127:
        if len(buf) < 10:
            return None
        if buf[2] & 0x80:  # 64-bit length MSB MUST be 0
            raise _ProtocolError(1002)
        # TODO(frame-size): enforce a configurable maximum frame length
        length = int.from_bytes(buf[2:10], "big")
        offset = 10
    if opcode >= 0x8 and (not fin or length > 125):
        raise _ProtocolError(1002)  # control frames: FIN=1, ≤125 bytes
    if len(buf) < offset + 4 + length:
        return None
    mask = buf[offset : offset + 4]
    raw = buf[offset + 4 : offset + 4 + length]
    payload = bytes(raw[i] ^ mask[i & 3] for i in range(length))
    return opcode, payload, offset + 4 + length


# --- Socket-like wrappers ---------------------------------------------------


class _PrependSocket:
    """The raw socket with ``prefix`` bytes served before it — used to push the
    sniffed first byte back on the raw-aiomsg path. Transparent otherwise: it
    forwards every socket operation the framing layer uses."""

    def __init__(self, prefix: bytes, raw):
        self._prefix = bytearray(prefix)
        self._raw = raw

    def recv_into(self, view) -> int:
        if self._prefix:
            n = min(len(view), len(self._prefix))
            view[:n] = self._prefix[:n]
            del self._prefix[:n]
            return n
        return self._raw.recv_into(view)

    def send(self, data) -> int:
        return self._raw.send(data)

    def settimeout(self, t):
        self._raw.settimeout(t)

    def fileno(self) -> int:
        return self._raw.fileno()

    def pending(self) -> bool:
        if self._prefix:
            return True
        p = getattr(self._raw, "pending", None)
        return bool(p and p())

    def close(self):
        self._raw.close()


class WSSocket:
    """Presents a WebSocket connection as a raw byte-stream socket.

    ``recv_into`` returns unmasked binary-message payload bytes (reassembling
    fragments, answering pings, dropping pongs, echoing close), and ``send``
    frames each write as one unmasked binary message. Persistent buffers survive
    the framing layer's short-timeout ``recv_into`` retries; ``pending`` lets the
    reader's readiness check (which selects on the fd) also see payload bytes
    already decoded inside this wrapper — bytes the fd's select can never show.
    """

    def __init__(self, raw, initial: bytes = b""):
        self._raw = raw
        self._raw_buf = bytearray(initial)  # bytes not yet parsed into frames
        self._out_buf = bytearray()  # decoded payload ready for recv_into
        self._eof = False
        self._close_sent = False

    # -- reading --
    def recv_into(self, view) -> int:
        while not self._out_buf:
            if self._eof:
                return 0
            self._pump()  # parse any complete frames already buffered
            if self._out_buf or self._eof:
                break
            chunk = self._raw.recv(65536)  # may raise TimeoutError (retried above)
            if not chunk:
                self._eof = True
                return 0
            self._raw_buf += chunk
        if self._eof and not self._out_buf:
            return 0
        n = min(len(view), len(self._out_buf))
        view[:n] = self._out_buf[:n]
        del self._out_buf[:n]
        return n

    def _pump(self):
        """Parse every complete frame currently buffered, handling control
        frames and appending binary payloads to ``_out_buf``."""
        while not self._eof:
            try:
                parsed = parse_client_frame(self._raw_buf)
            except _ProtocolError as e:
                self._send_close(e.code)
                self._eof = True
                return
            if parsed is None:
                return
            opcode, payload, consumed = parsed
            del self._raw_buf[:consumed]
            if opcode in (OP_BIN, OP_CONT):
                self._out_buf += payload
            elif opcode == OP_PING:
                self._send_raw(server_frame(OP_PONG, payload))
            elif opcode == OP_PONG:
                pass
            elif opcode == OP_TEXT:
                self._send_close(1003)  # text is not valid for aiomsg
                self._eof = True
                return
            elif opcode == OP_CLOSE:
                code = int.from_bytes(payload[:2], "big") if len(payload) >= 2 else 1000
                self._send_close(code)
                self._eof = True
                return
            else:
                self._send_close(1002)  # unknown opcode
                self._eof = True
                return

    # -- writing --
    def send(self, data) -> int:
        """Send ``data`` as one binary WS frame; returns len(data) (all of it)."""
        payload = bytes(data)
        self._send_raw(server_frame(OP_BIN, payload))
        return len(payload)

    def _send_raw(self, frame: bytes):
        """Write a full frame, retrying transient timeouts internally so a frame
        is never half-sent and re-encoded (which would corrupt the stream)."""
        mv = memoryview(frame)
        while mv:
            try:
                n = self._raw.send(mv)
            except _SEND_RETRYABLE:
                continue
            except OSError:
                return  # peer gone; teardown proceeds via recv EOF
            mv = mv[n:]

    def _send_close(self, code):
        if self._close_sent:
            return
        self._close_sent = True
        payload = code.to_bytes(2, "big") if code is not None else b""
        self._send_raw(server_frame(OP_CLOSE, payload))

    # -- socket surface --
    def settimeout(self, t):
        self._raw.settimeout(t)

    def fileno(self) -> int:
        return self._raw.fileno()

    def pending(self) -> bool:
        return bool(self._out_buf)

    def close(self):
        self._send_close(1000)
        self._raw.close()


# --- Entry point ------------------------------------------------------------


def _recv_until(raw, terminator: bytes, already: bytes):
    """Read from ``raw`` until ``terminator`` appears; return (before+term, rest)
    or raise WSError('400 …') if the request exceeds the cap or EOFs first."""
    buf = bytearray(already)
    while True:
        idx = buf.find(terminator)
        if idx != -1:
            end = idx + len(terminator)
            return bytes(buf[:end]), bytes(buf[end:])
        if len(buf) > MAX_REQUEST_BYTES:
            raise WSError("400 Bad Request")
        chunk = raw.recv(4096)
        if not chunk:
            raise WSError("400 Bad Request")
        buf += chunk


def sniff_and_wrap(conn):
    """Detect raw-vs-WebSocket on an accepted connection (PROTOCOL.md §10).

    Returns a socket-like object for the connection handler, or ``None`` if the
    connection should be closed (unknown first byte, EOF, timeout, or a rejected
    upgrade). ``conn`` is a plain or already-TLS-wrapped blocking socket.
    """
    conn.settimeout(HANDSHAKE_TIMEOUT)
    try:
        first = conn.recv(1)
    except (OSError, ssl.SSLError):
        return None
    if not first:
        return None

    if first == b"\x00":
        return _PrependSocket(first, conn)  # raw aiomsg; give the byte back
    if first != b"G":
        return None

    try:
        request, leftover = _recv_until(conn, b"\r\n\r\n", first)
        key = parse_upgrade(request)
    except WSError as e:
        try:
            conn.sendall(error_response(str(e)))
        except OSError:
            pass
        return None
    except (OSError, ssl.SSLError):
        return None

    try:
        conn.sendall(success_response(key))
    except OSError:
        return None
    return WSSocket(conn, initial=leftover)
